### PR TITLE
Implement client reg api

### PR DESCRIPTION
**Pull request by shevanfernando**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings

### CODE REVIEW

File: capstone-server/src/main/java/com/hcl/capstoneserver/config/GlobalErrorHandler.java [hunk 1]
```diff
@@ -85,5 +88,14 @@
                 HttpStatus.UNAUTHORIZED
         );
     }
+
+    @ExceptionHandler(DataIntegrityViolationException.class)
+    public final ResponseEntity<DefaultErrorResponse> handleConstraintViolatedErrors(Exception ex) {
```
Comment (ShashiSrinath) [inline capstone-server/src/main/java/com/hcl/capstoneserver/config/GlobalErrorHandler.java:93]:
```suggestion
    public final ResponseEntity<DefaultErrorResponse> handleConstraintViolatedErrors(DataIntegrityViolationException ex) {
```

File: capstone-server/src/main/java/com/hcl/capstoneserver/config/GlobalErrorHandler.java [hunk 2]
```diff
@@ -85,5 +88,14 @@
...
+        ex.printStackTrace();
```
Comment (ShashiSrinath) [inline capstone-server/src/main/java/com/hcl/capstoneserver/config/GlobalErrorHandler.java:94]:
```suggestion
      
```

don't need this. it is not an error

File: capstone-server/src/main/java/com/hcl/capstoneserver/config/GlobalErrorHandler.java [hunk 3]
```diff
@@ -62,14 +66,13 @@
     }
 
     @ExceptionHandler(BadCredentialsException.class)
-    public final ResponseEntity<DefaultErrorResponse> handleBadCredentials(BadCredentialsException ex) {
+    public final ResponseEntity<DefaultErrorResponse> handleBadCredentials() {
```
Comment (ShashiSrinath) [inline capstone-server/src/main/java/com/hcl/capstoneserver/config/GlobalErrorHandler.java:69]:
```suggestion
public final ResponseEntity<DefaultErrorResponse> handleBadCredentials(BadCredentialsException ex) {
```

keep the original structure

Comment (shevanfernando) [reply capstone-server/src/main/java/com/hcl/capstoneserver/config/GlobalErrorHandler.java:69]: But ex is never used.

File: capstone-server/src/main/java/com/hcl/capstoneserver/config/GlobalErrorHandler.java [hunk 4]
```diff
@@ -85,5 +90,32 @@
                 HttpStatus.UNAUTHORIZED
         );
     }
+
+    @ExceptionHandler(EmailAlreadyExistsException.class)
+    public final ResponseEntity<Object> handleEmailExistsException(EmailAlreadyExistsException ex) {
+        ArrayList<Map<String, String>> errors = errorsMap("email", ex.getMessage());
+        return new ResponseEntity<>(
+                new DefaultValidationErrorResponse(HttpStatus.BAD_REQUEST, "User Already Exists", errors),
```
Comment (ShashiSrinath) [inline capstone-server/src/main/java/com/hcl/capstoneserver/config/GlobalErrorHandler.java:98]: change the message to "An account with this email address already exists. please sign in or use another email"

File: capstone-server/src/main/java/com/hcl/capstoneserver/config/GlobalErrorHandler.java [hunk 5]
```diff
@@ -85,5 +90,32 @@
...
+    @ExceptionHandler(UserAlreadyExistsException.class)
+    protected final ResponseEntity<Object> handleUserIdExistsException(UserAlreadyExistsException ex) {
+        ArrayList<Map<String, String>> errors = errorsMap("userId", ex.getMessage());
+        return new ResponseEntity<>(
+                new DefaultValidationErrorResponse(HttpStatus.BAD_REQUEST, "User Already Exists", errors),
+                HttpStatus.BAD_REQUEST
+        );
+    }
+
+    private ArrayList<Map<String, String>> errorsMap(String field, String message) {
```
Comment (ShashiSrinath) [inline capstone-server/src/main/java/com/hcl/capstoneserver/config/GlobalErrorHandler.java:112]: if we are going to use a general method to create the error map, it should support generating multiple errors like the first method

**Review by ShashiSrinath — APPROVED** [review]

**Pull request merged**